core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,537 @@
+                  break;
+               case AVAILABLE:
+                  if (casState(LockState.AVAILABLE, LockState.ACQUIRED)) {
+                     return; //acquired!
+                  }
+                  break;
+               case ACQUIRED:
+                  return; //acquired!
+               case RELEASED:
+                  throw new IllegalStateException("Lock already released!");
+               case TIMED_OUT:
+                  cleanup();
+                  throw new TimeoutException("Timeout waiting for lock.");
+               case DEADLOCKED:
+                  cleanup();
+                  throw new DeadlockDetectedException("DeadLock detected");
+               default:
+                  throw new IllegalStateException("Unknown lock state: " + lockState);
+            }
+         }
+      }
+
+      @Override
+      public void addListener(LockListener listener) {
+         notifier.thenRun(() -> this.invoke(listener));
+      }
+
+      @Override
+      public void cancel(LockState state) {
+         checkValidCancelState(state);
+         out:
+         do {
+            LockState currentState = lockState;
+            switch (currentState) {
+               case WAITING:
+                  if (casState(LockState.WAITING, state)) {
+                     notifyListeners();
+                     break out;
+                  }
+                  break;
+               case ACQUIRED: //no-op, a thread is inside the critical section.
+               case TIMED_OUT:
+               case DEADLOCKED:
+               case RELEASED:
+                  return; //no-op, the lock is in final state.
+               default:
+                  if (casState(currentState, state)) {
+                     break out;
+                  }
+
+            }
+         } while (true);
+         onCanceled(this);
+      }
+
+      @Override
+      public Object getRequestor() {
+         return owner;
+      }
+
+      @Override
+      public Object getOwner() {
+         LockPlaceHolder owner = current;
+         return owner != null ? owner.owner : null;
+      }
+
+      @Override
+      public String toString() {
+         return "LockPlaceHolder{" +
+               "lockState=" + lockState +
+               ", owner=" + owner +
+               '}';
+      }
+
+      @Override
+      public void setDeadlockChecker(DeadlockChecker deadlockChecker) {
+         this.deadlockChecker = deadlockChecker;
+         LockPlaceHolder currentHolder = current;
+         if (currentHolder != null) {
+            checkDeadlock(currentHolder.owner);
+         }
+      }
+
+      private void invoke(LockListener invoker) {
+         LockState state = lockState;
+         switch (state) {
+            case WAITING:
+               throw new IllegalStateException("WAITING is not a valid state to invoke the listener");
+            case ACQUIRED:
+            case RELEASED:
+               invoker.onEvent(LockState.AVAILABLE);
```

Comment:
I'm ok with merging AVAILABLE and ACQUIRED completely, not just for the listener, but I'm not so sure about RELEASED. It seems to me like that one should be an error, like WAITING.